src/modules/keyboardmanager/test/BufferValidationTests.cpp

```diff
@@ -0,0 +1,3197 @@
+            Assert::AreEqual(true, result.second == BufferValidationHelpers::DropDownAction::NoAction);
+
+            // Case 17: Validate the element when making null-selection (-1 index) on third dropdown of first column of valid 3 key shortcut to shortcut row
+            result = BufferValidationHelpers::ValidateShortcutBufferElement(3, 0, 2, true, -1, std::vector<int32_t>({ (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_CONTROL)), (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_SHIFT)), -1 }), std::wstring(), keyList, remapBuffer, false);
+
+            // Assert that the element is valid and no drop down action is required
+            Assert::AreEqual(true, result.first == KeyboardManagerHelper::ErrorType::NoError);
+            Assert::AreEqual(true, result.second == BufferValidationHelpers::DropDownAction::NoAction);
+
+            // Case 18: Validate the element when making null-selection (-1 index) on third dropdown of second column of valid 3 key shortcut to shortcut row
+            result = BufferValidationHelpers::ValidateShortcutBufferElement(3, 1, 2, true, -1, std::vector<int32_t>({ (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_CONTROL)), (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_SHIFT)), -1 }), std::wstring(), keyList, remapBuffer, false);
+
+            // Assert that the element is valid and no drop down action is required
+            Assert::AreEqual(true, result.first == KeyboardManagerHelper::ErrorType::NoError);
+            Assert::AreEqual(true, result.second == BufferValidationHelpers::DropDownAction::NoAction);
+
+            // Case 19: Validate the element when making null-selection (-1 index) on third dropdown of second column of valid 3 key hybrid shortcut to shortcut row
+            result = BufferValidationHelpers::ValidateShortcutBufferElement(3, 1, 2, true, -1, std::vector<int32_t>({ (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_CONTROL)), (int32_t)std::distance(keyList.begin(), std::find(keyList.begin(), keyList.end(), VK_SHIFT)), -1 }), std::wstring(), keyList, remapBuffer, true);
+
+            // Assert that the element is valid and no drop down action is required
+            Assert::AreEqual(true, result.first == KeyboardManagerHelper::ErrorType::NoError);
+            Assert::AreEqual(true, result.second == BufferValidationHelpers::DropDownAction::NoAction);
+        }
+
+        // Test if the ValidateShortcutBufferElement method returns ShortcutStartWithModifier error and no drop down action is required on setting first drop down to an action key on a non-hybrid control column
+        TEST_METHOD (ValidateShortcutBufferElement_ShouldReturnShortcutStartWithModifierErrorAndNoAction_OnSettingFirstDropDownToActionKeyOnANonHybridColumn)
```

Comment:
discussion:  I personally find this test quite hard to read.   I have a couple suggestions, that I'd like your input on. 
1.  There are a log of parameters to the `ValidateShortcutBufferElement` potentially we could create an object to enapsulate these parametes, or alternatively pass named variables into the method when calling it. 
2.  My personal preference is to have one  assert (or cohesive group of asserts) per test.   If future developers need to refactor it's usually clearer what functionality has been broken this way, and you can be more confident that the tests are independent.

Reply:
Regarding 1, many of the arguments are indices or certain UI situations. Earlier these were directly extracted from the UI but for testing they were split into separate arguments. I'm not sure about encapsulating the params, but we could make a common method in the test class which takes the named variables and accordingly does the steps of adding the remap to the buffer as well as providing the arguments in the correct format for ValidateShortcutBufferElement.

Regarding 2, do you think it would be better to move each of them to a separate test? I essentially required the equivalent of the [TestCase] feature in C# unit tests, to ensure that all the different scenarios are handled, which is why for each test there are several act/assert sets (where each set is actually a test case). I need to keep the all those test cases to ensure all the scenarios are taken care of, but I wasn't sure on the best way to format it. Would it be better to have a separate test for each of those cases? Or is there something else I could leverage on the C++ testing framework to make it more generic? (Like having a single test method and calling it for different argument sets where each set is a test case).

Reply:
I've reduced the number of cases in several TEST_METHODs by adding tests for some of the smaller functions like `IsValidShortcut`, `DoKeysOverlap`, `GetKeysFromSelectedIndices` and `CheckRepeatedModifier`. Since MSTest (both the C++ and C# versions) don't have a a data driven test method, I created a helper RunTestCases, which takes the test case arguments and runs the test lambda over them. And for the test arguments for the Buffer Validation methods I've added structs to make them readable. Now all the tests for ValidateShortcutBufferElement (since they had multiple assert blocks) have been restructured as `Add test cases to test case vector`, followed by run test method on all test cases, where the test method has been structured as Arrange, Act, Assert. This should make it much more readable.

Reply:
I've also added type aliases to make the large type names more readable